frame/preimage/src/lib.rs

```diff
@@ -342,3 +388,74 @@ impl<T: Config> PreimageRecipient<T::Hash> for Pallet<T> {
 		debug_assert!(res.is_ok(), "unnote_preimage failed - request outstanding?");
 	}
 }
+
+/// Place temporary primages into non-persistent storage for use in the transaction's call.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct TempPreimages<T: Config + Send + Sync>(Vec<Vec<u8>>, sp_std::marker::PhantomData<T>)
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>;
+
+impl<T: Config + Send + Sync> sp_std::fmt::Debug for TempPreimages<T>
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>,
+{
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "TempPreimages")
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: Config + Send + Sync> TempPreimages<T>
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>,
+{
+	/// Create new `SignedExtension` to feed in preimages.
+	pub fn new(preimages: Vec<Vec<u8>>) -> Self {
+		Self(preimages, sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Config + Send + Sync> SignedExtension for TempPreimages<T>
+where
+	<T as frame_system::Config>::Call: IsSubType<Call<T>>,
+{
+	type AccountId = T::AccountId;
+	type Call = <T as frame_system::Config>::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	const IDENTIFIER: &'static str = "TempPreimages";
+
+	fn additional_signed(&self) -> Result<Self::AdditionalSigned, TransactionValidityError> {
+		Ok(())
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		Ok(self.validate(who, call, info, len).map(|_| ())?)
+	}
+
+	fn validate(
+		&self,
+		_who: &Self::AccountId,
+		_call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		for i in &self.0 {
+			TempPreimageFor::<T>::insert(T::Hashing::hash(&i[..]), i);
```

Comment:
I think if it should be possible to reference ephemeral hashes that other extrinsics provided, then the `requires` / `provides` tags are needed here, or?  
Otherwise the extrinsic ordering could be wrong.  

An enum like `MaybeHashed` could be used to create a dependency on a hash or a value of a previous extrinsic.